### PR TITLE
fixup! Introduce caching cost and stats calculator

### PR DESCRIPTION
**Pull request by kokosing**
fixup! Introduce caching cost and stats calculator

Cummulative cost cannot be used to calculate regular cost as it causes
cost to sky rocket!

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/cost/CostCalculatorUsingExchanges.java [hunk 1]
```diff
@@ -108,13 +116,29 @@ public PlanNodeCostEstimate visitOutput(OutputNode node, Void context)
         @Override
         public PlanNodeCostEstimate visitFilter(FilterNode node, Void context)
         {
-            return cpuCost(getStats(node.getSource()).getOutputSizeInBytes());
+            // not accessed columns does not cost
+            Set<Symbol> usedSymbols = DependencyExtractor.extractUnique(node.getPredicate());
+            Set<Symbol> notUsedSymbols = Sets.difference(new HashSet<>(node.getSource().getOutputSymbols()), usedSymbols);
+
+            PlanNodeStatsEstimate filterStats = getStats(node.getSource());
+            PlanNodeStatsEstimate filteredFilterStats = PlanNodeStatsEstimate.buildFrom(filterStats).removeSymbolStatistics(notUsedSymbols).build();
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/cost/CostCalculatorUsingExchanges.java]: consider `s/filteredFilterStats/onlyUsedSymbolsStats/`

Why unused columns do not cost? Are we always producing dictionary blocks in filter operator?

File: presto-main/src/main/java/com/facebook/presto/cost/CostCalculatorUsingExchanges.java [hunk 2]
```diff
@@ -108,13 +116,29 @@ public PlanNodeCostEstimate visitOutput(OutputNode node, Void context)
         @Override
         public PlanNodeCostEstimate visitFilter(FilterNode node, Void context)
         {
-            return cpuCost(getStats(node.getSource()).getOutputSizeInBytes());
+            // not accessed columns does not cost
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/cost/CostCalculatorUsingExchanges.java]: `s/does/do`

File: presto-main/src/main/java/com/facebook/presto/cost/CostCalculatorUsingExchanges.java [hunk 3]
```diff
@@ -108,13 +116,29 @@ public PlanNodeCostEstimate visitOutput(OutputNode node, Void context)
...
+            PlanNodeStatsEstimate filterStats = getStats(node.getSource());
+            PlanNodeStatsEstimate filteredFilterStats = PlanNodeStatsEstimate.buildFrom(filterStats).removeSymbolStatistics(notUsedSymbols).build();
+            return cpuCost(filteredFilterStats.getOutputSizeInBytes());
         }
 
         @Override
         public PlanNodeCostEstimate visitProject(ProjectNode node, Void context)
         {
-            return cpuCost(getStats(node).getOutputSizeInBytes());
+            // literal or identity projection does not cost
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/cost/CostCalculatorUsingExchanges.java]: `s/does/do`

File: presto-main/src/main/java/com/facebook/presto/cost/CostCalculatorUsingExchanges.java [hunk 4]
```diff
@@ -108,13 +116,29 @@ public PlanNodeCostEstimate visitOutput(OutputNode node, Void context)
...
+            PlanNodeStatsEstimate projectStats = getStats(node);
+            Set<Symbol> notUsedSymbols = node.getAssignments().getSymbols().stream()
+                    .filter(symbol -> {
+                        Expression expression = node.getAssignments().get(symbol);
+                        return expression instanceof SymbolReference || expression instanceof Literal;
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/cost/CostCalculatorUsingExchanges.java:136]: Why literal doesn't cost? Don't we need to put those values into blocks?

File: presto-main/src/main/java/com/facebook/presto/cost/CostCalculatorUsingExchanges.java [hunk 5]
```diff
@@ -169,7 +193,9 @@ public PlanNodeCostEstimate visitExchange(ExchangeNode node, Void context)
         @Override
         public PlanNodeCostEstimate visitTableScan(TableScanNode node, Void context)
         {
-            return cpuCost(getStats(node).getOutputSizeInBytes()); // TODO: add network cost, based on input size in bytes?
+            // memory to cpu througput is about 10GB/s
+            // disk to memory is 100MB/s - 100 slower
+            return cpuCost(getStats(node).getOutputSizeInBytes() * 100); // TODO: add network cost, based on input size in bytes?
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/cost/CostCalculatorUsingExchanges.java]: This assumes you read from disk, which may or may not be the case. Cost of reading is something a connector should be able to provide (consider adding relevant TODO in the code). 

Even if you assume we read from HDFS, this is both CPU cost and network cost.

What about assuming this is `(cpu=dataSize, mem=0, net=dataSize)` for now?

File: presto-main/src/main/java/com/facebook/presto/cost/PlanNodeStatsEstimate.java [hunk 6]
```diff
@@ -172,6 +173,14 @@ public Builder addSymbolStatistics(Map<Symbol, SymbolStatsEstimate> symbolStatis
             return this;
         }
 
+        public Builder removeSymbolStatistics(Collection<Symbol> symbols)
+        {
+            for (Symbol symbol : symbols) {
+                removeSymbolStatistics(symbol);
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/cost/PlanNodeStatsEstimate.java]: I wonder whether directly using `minusAll(..)` would offer better performance?

File: presto-main/src/main/java/com/facebook/presto/cost/CachingCostCalculator.java [hunk 7]
```diff
@@ -42,7 +42,7 @@ public PlanNodeCostEstimate calculateCost(PlanNode planNode, Lookup lookup, Sess
     {
         if (!costs.containsKey(planNode)) {
             // cannot use Map.computeIfAbsent due to costs map modification in the mappingFunction callback
-            PlanNodeCostEstimate cost = costCalculator.calculateCumulativeCost(planNode, lookup, session, types);
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/cost/CachingCostCalculator.java:45]: Maybe `calculateCumulativeCost` should be cached too?

**Review by findepi — CHANGES_REQUESTED** [review]

File: presto-main/src/main/java/com/facebook/presto/cost/CostCalculatorUsingExchanges.java (hunk 4 above)
Comment (kokosing) [reply presto-main/src/main/java/com/facebook/presto/cost/CostCalculatorUsingExchanges.java:136]: no, AFAIR dictionary block is used

File: presto-main/src/main/java/com/facebook/presto/cost/CostCalculatorUsingExchanges.java (hunk 5 above)
Comment (kokosing) [reply presto-main/src/main/java/com/facebook/presto/cost/CostCalculatorUsingExchanges.java]: Yes, yes, yes. This is why I wrote that formula is not ideal.

I wanted to be make this a cost higher than the cost of projection.

File: presto-main/src/main/java/com/facebook/presto/cost/CachingCostCalculator.java (hunk 7 above)
Comment (kokosing) [reply presto-main/src/main/java/com/facebook/presto/cost/CachingCostCalculator.java:45]: Yes. Moreover this class needs tests.

File: presto-main/src/main/java/com/facebook/presto/cost/CostCalculatorUsingExchanges.java (hunk 1 above)
Comment (kokosing) [reply presto-main/src/main/java/com/facebook/presto/cost/CostCalculatorUsingExchanges.java]: I thought so, but I could be wrong. I can remove this formula.

**Comment by kokosing** [comment]
@findepi comments applied

File: presto-main/src/main/java/com/facebook/presto/cost/CostCalculatorUsingExchanges.java (hunk 5 above)
Comment (findepi) [reply presto-main/src/main/java/com/facebook/presto/cost/CostCalculatorUsingExchanges.java]: as disc. offline, bump network cost and consider using smaller multiplier (10?)

**Review by findepi — APPROVED** [review]

**Comment by findepi** [comment]
one minor comment

**Pull request merged**